Allow for customisation of chain selection systems

gnunicorn opened the pull request:
Infrastructure ref #1125 . Superseeds #1443.

This breaks off the `SelectChain` into its own trait, externalised from `client`, in `common-consensus` and then provides the current implementation as `LongestChain` for convenience. In order to allow others to specify a custom chain, this now also needs to be handled separately in the service definition and custom setups for the consensus defined in there. Thus, this breaks compatibility with existing implementations.

The general infrastructure is there, but this PR still has a few issues that needs resolving before it can be merged. However, reviewing (especially around the architecture in general) can already proceed.

The known issues to fix are:
 - [x] Update Tests
 - [x] several FIXMEs were added and need to be either documented or resolved
 - ~~most likely resolving of conflicts once #2165 is merged~~

bkchr commented on core/consensus/aura/slots/src/lib.rs:
```diff
@@ -120,14 +119,14 @@ pub fn start_slot_worker_thread<B, C, W, SO, SC, OnExit>(
 pub fn start_slot_worker<B, C, W, SO, SC, OnExit>(
 	slot_duration: SlotDuration,
-	client: Arc<C>,
+	client:  C,
```
```suggestion
	client: C,
```

bkchr commented on core/consensus/aura/slots/src/lib.rs:
```diff
@@ -120,14 +119,14 @@ pub fn start_slot_worker_thread<B, C, W, SO, SC, OnExit>(
 ) -> Result<impl Future<Item=(), Error=()>, consensus_common::Error> where
 	B: Block,
-	C: ChainHead<B>,
+	C: SelectChain<B> +  Clone,
```
```suggestion
	C: SelectChain<B> + Clone,
```

bkchr commented on core/consensus/common/src/select_chain.rs:
```diff
@@ -0,0 +1,86 @@
+
```
```suggestion
```

bkchr commented on core/consensus/common/src/select_chain.rs:
```diff
@@ -0,0 +1,86 @@
+
+	/// Get best block header for authoring
+	fn best_block_header_for_authoring(&self) -> Result<<Block as BlockT>::Header, Error> {
+        self.best_block_header()
```
Spaces.

bkchr commented on core/consensus/common/src/select_chain.rs:
```diff
@@ -0,0 +1,86 @@
+
+	/// Get best block header for finalisation
+	fn best_block_header_for_finalisation(&self) -> Result<<Block as BlockT>::Header, Error> {
+        self.best_block_header()
```
Spaces.

rphmeier commented on core/consensus/aura/src/lib.rs:
```diff
@@ -214,10 +215,11 @@ pub fn start_aura_thread<B, C, E, I, P, SO, Error, OnExit>(
 	slot_duration: SlotDuration,
 	local_key: Arc<P>,
 	client: Arc<C>,
+	select_chain: SC,
```
unfortunately, this breaks API. We will have to find a way to do this that doesn't.

gnunicorn replied:
Yes, this change will inevitably break the api - as it needs to be specified and at the service-intend-layer and somehow pushed down to the authoring and finalisation. Even if we tried to still have client facilitate this somehow, this would mean adding the same `SC` on every `Where C: Client<...SC>`, which would also break this same call.

I am afraid there is just no way around this breaking the API.

gnunicorn replied:
The only thing we _could_  do is have client implement the new trait as well and hard-code that to the `LongestChain` implementation and directly deprecate its usage. But then still, would the where clause later on have to change to the new `trait` and thus, this would still break the API.

gnunicorn replied:
This is currently the blocker for me on this PR. I am not sure how we even can resolve this without creating a huge other mess elsewhere.

The _only_ way I can currently think of is by having a compatibility layer, that is `Client` implements `LongestChain` `SC`, however in that case, a) the where clause still needs to be extended (and that is "breaking the API", as it now constrains more than it was before) and b) it would mean this function doesn't actually allow any other implementation and thus isn't really implementing the ticket (or solving our problem).

Not sure what to do about this?!?

rphmeier commented on core/consensus/common/src/select_chain.rs:
```diff
@@ -0,0 +1,86 @@
+	fn best_block_header(&self) -> Result<<Block as BlockT>::Header, Error>;
+
+	/// Get best block header for authoring
+	fn best_block_header_for_authoring(&self) -> Result<<Block as BlockT>::Header, Error> {
```
docs should have some distinction about what this means

gnunicorn replied:
I clarified the interface a bit and added docs to them explaining the intended usage

rphmeier commented on core/consensus/common/src/select_chain.rs:
```diff
@@ -0,0 +1,86 @@
+    }
+
+	/// Get best block header for finalisation
+	fn best_block_header_for_finalisation(&self) -> Result<<Block as BlockT>::Header, Error> {
```
should an instant-finality authoring system use this or `_for_authoring`?

gnunicorn replied:
Good question. Was also wondering whether `_finalisation` should fallback to `_authoring` in the main trait to make this easier...

rphmeier replied:
We might want to express it as a subset relationship. The `best_for_authoring` must always be equal to or a descendent of the `best_for_finalization`

DemiMarie-parity replied:
@rphmeier I think anything that does not support swappable consensus is a non-starter.  Ideally, consensus could be implemented in the runtime.

rphmeier commented on core/consensus/common/src/select_chain.rs:
```diff
@@ -0,0 +1,86 @@
+        self.best_block_header()
+    }
+
+	/// Get the most recent block hash of the best chain that contain block
```
double indent

gnunicorn commented on core/service/src/components.rs:
```diff
@@ -396,13 +406,21 @@ pub trait Components: Sized + 'static {
 /// A struct that implement `Components` for the full client.
 pub struct FullComponents<Factory: ServiceFactory> {
-	_factory: PhantomData<Factory>,
+	// _factory: PhantomData<Factory>,
```
this looks like it can be removed entirely.

gnunicorn commented on core/consensus/common/Cargo.toml:
```diff
@@ -6,6 +6,7 @@ description = "Common utilities for substrate consensus"
 edition = "2018"
 
 [dependencies]
+parking_lot = "0.7.1"
```
unnecessary.
```suggestion
```

gnunicorn commented on core/service/src/components.rs:
```diff
@@ -550,17 +577,30 @@ impl<Factory: ServiceFactory> Components for LightComponents<Factory> {
+		_config: &mut FactoryFullConfiguration<Self::Factory>,
+		_client: Arc<ComponentClient<Self>>
+	) -> Result<Self::SelectChain, error::Error> {
+		// FIXME: this (and other places) need a "NotAvailableInMode" for never-in-light-client-features
```
Not a problem rn.
```suggestion
```

gnunicorn commented on core/service/src/components.rs:
```diff
@@ -550,17 +577,30 @@ impl<Factory: ServiceFactory> Components for LightComponents<Factory> {
+		_client: Arc<ComponentClient<Self>>
+	) -> Result<Self::SelectChain, error::Error> {
+		// FIXME: this (and other places) need a "NotAvailableInMode" for never-in-light-client-features
+		// that doesn't break creation of light clients  
```
```suggestion
```

rphmeier commented on core/client/src/client.rs:
```diff
@@ -1529,12 +1529,14 @@ where
 	Block: BlockT<Hash=H256>,
 {
-	fn best_block_header(&self) -> Result<<Block as BlockT>::Header, ConsensusError> {
+	fn best_block_header_for_authoring(&self)
```
I'm not a big fan of this distinction. what does it mean to be "best for authoring" vs. best in general?

gnunicorn replied:
Me neither, it is only coming from the pragmatic problem, you mentioned, where we'd have two distinct ways (potentially coming to different outcomes) to determine what the `best_block` is based upon whether we finalise or we propose in polkadot. Or did I misunderstand that?

rphmeier replied:
yes, that is one practical problem we have. I think we ought to have an abstraction that covers it and a clear definition of what the separation is.

rphmeier commented on core/consensus/common/src/select_chain.rs:
```diff
@@ -0,0 +1,66 @@
+	fn best_block_header_for_authoring(&self) -> Result<<Block as BlockT>::Header, Error>;
+
+	/// Get best block header for finalisation
+	fn best_block_header_for_finalisation(&self) -> Result<<Block as BlockT>::Header, Error> {
```
btw. american spelling is used in the codebase so it should be `best_block_header_for_finalization`. Although I suggest shortening to `authoring_head` and `finalization_head`.

DemiMarie-parity commented on core/client/src/client.rs:
```diff
@@ -1477,21 +1355,202 @@ where
+/// where 'longest' is defined as the highest number of blocks
+pub struct LongestChain<B, Block> {
+	backend: Arc<B>,
+	import_lock: Arc<Mutex<()>>,
```
Given that `()` is obviously `Send` and `Sync`, there should be a comment here explaining what exactly the `Mutex` is protecting.  Even better would be to replace `()` by the actual type being protected.

gnunicorn replied:
This is an ugly hack from back in the day when we split importing out of the client and all of the sudden we could do things in parallel with importing that we didn't want to do while importing. Before, the client would hold the "currently importing hash" but that wasn't possible anymore after, so we created a lock that is acquired when importing takes place -- and this particular implementation of longest chain needs to acquire that lock to work (unfortunately). As my focus was on the chain selection, I didn't attempt to fix this ugliness, but IMHO the lock should actually be in the `ImportQueue` and something you coordinate with it, rather than sharing the lock here.

gnunicorn replied:
I'd put this into a ticket once we got this merged. Might be a good starter task.

DemiMarie-parity commented on core/client/src/client.rs:
```diff
@@ -1477,21 +1355,202 @@ where
+			Err(e) => return Err(error::Error::from_blockchain(Box::new(e)))
+		};
+		Ok(self.backend.blockchain().header(BlockId::Hash(info.best_hash))?
+			.expect("Best block header must always exist"))
```
It would be good to have a test case that checks that this is true, especially on genesis.

gnunicorn replied:
I was looking into this, but I don't think I want to introduce this in this PR.

As genesis is chain-specific, this test could be done against our test-client (which is a non-test) or as an integration test on `node` or other implementations. However, for the later, we don't really have any of these tests now, nor have we thought thoroughly enough how we want to support for, implement and run these tests (for e.g. I'd love to have a bunch of test in the template, that ensure assumptions like this for the genesis and service config developer might write and give them hints as to how to fix things if the test complain).

Adding this specific tests would require a bigger degree of changes in `node`/`node-template` to facilitate for this, but I don't think we have thought well enough of what we want there to add this hastily at this point.

DemiMarie-parity commented on core/client/src/client.rs:
```diff
@@ -1477,21 +1355,202 @@ where
+		};
+
+		// for each chain. longest chain first. shortest last
+		for leaf_hash in leaves {
```
This looks like it could be quadratic.

rphmeier replied:
Do you have a better algorithm?

DemiMarie-parity replied:
Not really, but we should make sure that this cannot be used in a denial-of-service attack.

gnunicorn replied:
This PR focuses on splitting out the existing algorithm. I am fine with fixing it, but that is out of scope of this PR.

rphmeier commented on core/client/src/client.rs:
```diff
@@ -1715,8 +1774,14 @@ pub(crate) mod tests {
+
 
-		assert_eq!(genesis_hash.clone(), client.best_containing(genesis_hash.clone(), None).unwrap().unwrap());
+		assert_eq!(genesis_hash.clone(), longest_chain_select.best_containing_for_authoring(
```
@DemiMarie-parity is this not the test that you were asking for?

DemiMarie-parity replied:
It is, I just didn’t see it.

gnunicorn commented on core/client/src/client.rs:
```diff
@@ -320,10 +311,23 @@ impl<B, E, Block, RA> Client<B, E, Block, RA> where
 
 	/// Expose backend reference. To be used in tests only
+	#[doc(hidden)]
+	#[deprecated(since="1.0.1", note="Rather than relying on `client` \
```
Not sure, it existed before this PR (and thus before it was deprecated officially, though the docs stated it), however it wasn't released ever before, so I see your argument. I am fine with it either way, if you prefer I removed it, I gladly do so.

rphmeier commented:
re: backwards compatibility. I think it's fine to break compatibility on this branch, although it may make backporting consensus changes to 1.0 more difficult.

gnunicorn commented on core/finality-grandpa/src/lib.rs:
```diff
@@ -52,6 +52,9 @@
 //! or prune any signaled changes based on whether the signaling block is
 //! included in the newly-finalized chain.
 #![forbid(warnings)]
+/// FIXME
```
I'll happily put this into a ticket, once review is done.

(Why is this here: because I officially deprecated the usage of `client.backend()` and the `forbid` then refuses to compile this file. However the deprecation, nor the changes needed in this file are strictly part of this PR and can be done as a nice starter-ticket)

gnunicorn replied:
Ticket now at https://github.com/paritytech/substrate/issues/2532

gnunicorn commented:
I've updated the PR to latest master (including changes to `babe`) and changed the interface as we've discussed, rather than having the weird four methods, we split up on a generic `best_chain` (primarily for authoring) but also let `SelectChain` choose a different block for finalization (which we will need for Polkadot). Makes the entire interface a bit cleaner.

Also addressed the other comments. Please review again, @DemiMarie-parity, @rphmeier

gavofyork commented:
*bump* @DemiMarie-parity @rphmeier

rphmeier commented on core/consensus/aura/src/lib.rs:
```diff
@@ -916,6 +918,10 @@ mod tests {
 			let client = net.lock().peer(*peer_id).client().clone();
+			let select_chain = LongestChain::new(
+				client.backend().clone(),
+				client.import_lock().clone()
```
```suggestion
				client.import_lock().clone(),
```

rphmeier commented on core/consensus/babe/src/lib.rs:
```diff
@@ -847,10 +851,11 @@ fn claim_slot(
 
 #[cfg(test)]
-#[allow(dead_code, unused_imports)]
+#[allow(dead_code, unused_imports, deprecated)] //FIXME
```
what's deprecated here? does it need an issue?

gnunicorn replied:
same reason. marked it with the issue

rphmeier commented on core/consensus/common/src/select_chain.rs:
```diff
@@ -0,0 +1,54 @@
+/// The SelectChain trait defines the strategy upon which the head is chosen
+/// if multiple forks are present for an opaque definition of "best" in the 
+/// specific chain build.
+////
```
```suggestion
///
```

rphmeier commented on core/consensus/common/src/select_chain.rs:
```diff
@@ -0,0 +1,54 @@
+	fn best_chain(&self) -> Result<<Block as BlockT>::Header, Error>;
+
+	/// Get the most recent block hash of the best chain that contain block
+	/// with the given `target_hash` for finalisation
```
This is a break from what we discussed while designing, which was that `finality_target` would find an _ancestor_ of `target_hash`. So to find a block to finalize, you do `finality_target(best_chain())`. For a lot of chain selection systems, `finality_target` is the identity function, but for some we might not be able to finalize all the blocks in that chain, so it would look backwards a bit.

gnunicorn replied:
I must have mixed something up then, because I thought this was the API we discussed. I think I just copied the docs over and didn't update them properly. But the idea was that given `best_chain()` (which is what most finality gadget will use prior) the SC would pick the most recent block that can be/should be finalized.

I'll update the doc to reflect that.

gnunicorn replied:
@rphmeier I changed the docs to reflect closer what we had in mind. Please confirm, that this the API we had discussed.

rphmeier replied:
@gnunicorn This is a good point. I think the only solution is to specify that the `finality_target` is the _block we would like to see finalized_, which excludes blocks that are already finalized. So a `finality_target` should be able to return `None` to indicate that there is no viable chain to finalize.

rphmeier commented on core/finality-grandpa/src/environment.rs:
```diff
@@ -317,7 +324,7 @@ impl<Block: BlockT<Hash=H256>, B, E, N, RA> grandpa::Chain<Block::Hash, NumberFo
 		debug!(target: "afg", "Finding best chain containing block {:?} with number limit {:?}", block, limit);
 
-		match self.inner.best_containing(block, None) {
+		match self.select_chain.finality_target(block, None) {
```
yeah, I guess the main thing is that we need a way to get the best chain containing a given hash. I thought of doing it as a separate operation, where we would use `sc.finality_target(sc.best_containing(included_hash))`

rphmeier replied:
I think the API in this PR is workable for GRANDPA, but I think we need to introduce a split if we want something that would work for finality gadgets and instant-finality authorship and probablistic finality authorship. 

With the current API I'm not sure exactly what a `Tendermint` implementation would build on.

gnunicorn replied:
I think this was mostly because of the same (wrong) API docs. I updated those.

The way I'd see this work is, that if you have a finality gadget, you'd pass the `best_chain` to `finality_target` and it has some way to find the ancestor that you should vote on.

For instant finality systems (like Tendermint or Rhododendron) I don't see why you'd ever need to use `finality_target` (or any finality gadget for that matter) at all. Because `best_chain` will already return a finalised block, thus the default implementation just returns the same hash again. Same for any probabilistic finality system but the other way around - no block is ever "surely finalized" - thus the system would not use the `finality_target` function - or given a hash and a certain risk profile could return some ancestor it is certain to not be reverted, but as there won't be any finality gadget, no one would ask for it...

rphmeier replied:
@gnunicorn I think if we would apply e.g. the polkadot chain-selection rule to an instant-finality authorship system, the behavior we'd desire is that
  - we have a best block which is finalized 
  - but we refuse to propose or vote on a new block until we've received the erasure-coded chunks for that chain. This would be accomplished by doing `finality_target(best_chain()) == best_chain()`.

rphmeier commented on core/finality-grandpa/src/tests.rs:
```diff
@@ -410,6 +418,11 @@ fn run_to_completion_with<F>(
 		};
 
+		// let select_chain = LongestChain::new(
+		// 	client.backend().clone(),
```
commented code here
